src/main/java/io/openliberty/tools/intellij/util/DebugModeHandler.java

```diff
@@ -0,0 +1,278 @@
+            ExecutionUtil.runConfiguration(settings, DefaultDebugExecutor.getDebugExecutorInstance(), DefaultExecutionTarget.INSTANCE, groupId);
+        } catch (Exception e) {
+            // do not show error if debug attachment was cancelled by user
+            if (!(e instanceof ProcessCanceledException)) {
+                LOGGER.error(String.format("Cannot connect debugger to port %d", debugPort), e);
+                ApplicationManager.getApplication().invokeLater(() -> Messages.showErrorDialog(LocalizedResourceUtil.getMessage("cannot.connect.debug.port", debugPort), "Liberty"));
+            }
+        }
+    }
+
+    /**
+     * Waits for the JDWP socket on the JVM to start listening for connections
+     *
+     * @param monitor progress monitor
+     * @param libertyModule Liberty module
+     * @param host JVM host to connect to
+     * @param debugPort JVM port to connect to
+     * @return debug port as a String
+     * @throws Exception
+     */
+    private String waitForSocketActivation(ProgressIndicator monitor, LibertyModule libertyModule, String host, int debugPort) throws Exception {
+        byte[] handshakeString = "JDWP-Handshake".getBytes(StandardCharsets.US_ASCII);
+        int retryLimit = 120;
+        int envReadMinLimit = 30;
+        int envReadMaxLimit = 60;
+        int envReadInterval = 5;
+
+        Path serverEnvPath = getServerEnvPath(libertyModule);
+
+        for (int retryCount = 0; retryCount < retryLimit; retryCount++) {
+            // check if cancelled
+            if (monitor.isCanceled()) {
+                return null;
+            }
+
+            if (serverEnvPath == null) {
+                serverEnvPath = getServerEnvPath(libertyModule);
+                TimeUnit.SECONDS.sleep(1);
+                continue;
+            }
+
+            // Check the server.env at the default location for any updates to WLP_DEBUG_ADDRESS for any changes.
+            // There is a small window in which the allocated port could have been taken by another process.
+            // If the port is already in use, dev mode will allocate a random port and reflect that by updating the server.env file.
+            if (retryCount >= envReadMinLimit && retryCount < envReadMaxLimit && (retryCount % envReadInterval == 0)) {
+                String envPortStr = readDebugPortFromServerEnv(serverEnvPath.toFile());
+                if (envPortStr != null) {
+                    int envPort = Integer.parseInt(envPortStr);
+                    if (envPort != debugPort) {
+                        debugPort = envPort;
+                    }
+                }
+            }
+
+            try (Socket socket = new Socket(host, debugPort)) {
+                socket.getOutputStream().write(handshakeString);
+                return String.valueOf(debugPort);
+            } catch (ConnectException e) {
+                TimeUnit.SECONDS.sleep(1);
```

Comment:
I didn't get why we need to wait for 1s inside the catch if it seems that is not trying to write the handshake again. Is that 1s wait necessary?